#69

javier-gomez-hs opened the pull request:
- Added new "withAttempts" parameter to the getPayments method in the Resource
- If it is true, added changes in the PaymentProcessor to get and handle the next Scheduled Payment Attempt
- Added changes to tests, Json and Resource files that includes this new attribute.

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -462,7 +471,54 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+                                  );
+    }
+
+    private PaymentAttempt getNextRetryDateForFailedPayment(final List<PaymentTransaction> purchasedTransactions) {
```
We should not recompute the payment retry dates but instead fetch the future notifications associated with these pending notifications. You will need to use the notification apis to access that information:
* A code example to fetch such future notification can be found [here](https://github.com/killbill/killbill/blob/master/overdue/src/main/java/org/killbill/billing/overdue/notification/DefaultOverduePosterBase.java#L110)
* For context, those notifications are inserted [here](https://github.com/killbill/killbill/blob/master/payment/src/main/java/org/killbill/billing/payment/retry/BaseRetryService.java#L116). This will give you a clue on which notification `service` and `queueName` to use and what they contain.

sbrossie commented:
@pierre and I have discussed the changes (and Pierre is commenting on the `killbill-api` PR#15). This will have a significant impact on your code change. I only commented on the piece that fetches the future payment retry notification since that will need to be implemented anyways.

@javier-gomez-hs The task is not as trivial as it seems, so i expect a few back and forth before you get it right, but this will be a great learning in the end.

javier-gomez-hs commented:
@sbrossie @pierre 
I've added the requested changes.
Please take a look and feel free to add your comments!

Thanks.

javier-gomez-hs commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -472,53 +483,53 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+                        ((PaymentRetryNotificationKey) notificationEvent.getEvent()).getPaymentControlPluginNames().get(0), //pluginName,
+                        null);//pluginProperties
+                result.add(futurePaymentAttempt);
             }
```
@sbrossie @pierre 
This is all the information I have to build the future Payment Attempt from the Notification Event in the queue.
Should I complete all the attributes with one of the past Attempts? Or the Transactions?

sbrossie replied:
Yes, you should complete based on previous attempt. Some fields will be left null (as they are not yet known)
* `paymentMethodId`
* `transactionId`
* `createdDate`/`updatedDate` (but we should set `effectiveDate`)

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -472,53 +483,53 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+        // Get Future Payment Attempts from Notification Queue and add them to the list
+        try {
+            final NotificationQueue retryQueue = notificationQueueService.getNotificationQueue(DefaultPaymentService.SERVICE_NAME, "retry");
+            final List<NotificationEventWithMetadata<NotificationEvent>> notificationEventWithMetadatas =
```
Use the constant `DefaultRetryService.QUEUE_NAME` (instead of "retry")

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -491,45 +491,66 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+                    pastPaymentAttempt.getAmount(),
+                    pastPaymentAttempt.getCurrency(),
+                    pastPaymentAttempt.getPluginName(),
+                    (lastPaymentTransaction.getPaymentInfoPlugin() != null) ? lastPaymentTransaction.getPaymentInfoPlugin().getProperties() : null);
```
Why are you using the `lastPaymentTransaction` to fill in the plugin properties associated with all past attempts? Instead you need to deserialize the `pluginProperties` associated with each attempt.

javier-gomez-hs replied:
Fixed

sbrossie replied:
Not sure if this is GH issue, but i am not seeing any change ?

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -491,45 +491,66 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
-                        null);//pluginProperties
-                result.add(futurePaymentAttempt);
+                        lastPaymentAttemptModelDao.getAccountId(), // accountId
+                        lastPaymentAttemptModelDao.getPaymentMethodId(), // paymentMethodId
```
The `lastPaymentAttemptModelDao` is **not necessarily** the one attached to the `futurePaymentAttempt`. Instead you need to find the most recent attempt from `pastPaymentAttempts` that matches your `paymentAttemptId`

javier-gomez-hs replied:
So you mean getting all the attempts that matches the **paymentAttemptId** and then find the one that has the most recent **createdDate** or **updatedDate**?

sbrossie replied:
Yep something like that. I would not look at `createdDate` or `updatedDate`, but i would iterate over `pastPaymentAttempts ` starting from last one (DB query already orders them), and stopping when i find the first match (based on `paymentAttemptId`). Actually i am not even sure this is necessary to go reverse order because there should be only one match, but just for safety maybe.

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -491,45 +491,66 @@ public boolean apply(final PaymentTransactionModelDao curPaymentTransactionModel
+                        lastPaymentAttemptModelDao.getTransactionType(), // transactionType
+                        SCHEDULED, // stateName
+                        lastPaymentTransaction.getAmount(), // amount
+                        lastPaymentTransaction.getCurrency(), // currency
```
Same here, using `lastPaymentTransaction` may not work.

javier-gomez-hs replied:
Fixed

sbrossie replied:
Same here: Not sure if this is GH issue, but i am not seeing any change ?

javier-gomez-hs commented:
@sbrossie @pierre 
I've just pushed my latest changes to this PR and the [API](https://github.com/killbill/killbill-api/pull/15) and [Java Client](https://github.com/killbill/killbill-client-java/pull/39)

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java:
```diff
@@ -584,15 +605,103 @@ public int compare(final PaymentTransaction o1, final PaymentTransaction o2) {
+                }
+            }
+        }
+        return null;
```
Nit: That should never happen so there is a question whether we should throw or return null

javier-gomez-hs replied:
Well..I'm making sure [here](https://github.com/javier-gomez-hs/killbill/blob/12af0bf36ba5b2e98504987b667c3876223aaaaf/payment/src/main/java/org/killbill/billing/payment/core/PaymentProcessor.java#L659-L659) that nothing will happen if the Last Payment Attempt is **null**. It will just not take into account the Future Attempts, which sounds logic.
Does that sounds good for you guys too?

sbrossie commented:
Looks great.

The only i would add is a jaxrs test (using java client ) to fetch payment with attempts. You could trigger payment failures, example [here](https://github.com/killbill/killbill/blob/master/profiles/killbill/src/test/java/org/killbill/billing/jaxrs/TestPayment.java#L102), and verify we see the `SCHEDULED` attempt and it correctly populated.

javier-gomez-hs commented:
@sbrossie @pierre 
I've added a jaxrs test to verify the new behavior.

Please feel free to add any comment / suggestion.

Thanks!

sbrossie commented:
👍  I think this is ready to be merged (i'll delay it for a bit as we have other work in parallel).

javier-gomez-hs commented:
@pierre @sbrossie 
I've updated all the classes and integration tests that are using the PaymentApi methods since they were updated with the new **withAttempts** property.

Thanks.

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -722,31 +722,31 @@ public Payment createChargebackReversal(final Account account, final UUID paymen
     @Override
-    public List<Payment> getAccountPayments(final UUID accountId, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext tenantContext) throws PaymentApiException {
+    public List<Payment> getAccountPayments(final UUID accountId, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext tenantContext) throws PaymentApiException {
         return paymentProcessor.getAccountPayments(accountId, withPluginInfo, tenantContext, internalCallContextFactory.createInternalTenantContext(accountId, tenantContext));
```
The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -722,31 +722,31 @@ public Payment createChargebackReversal(final Account account, final UUID paymen
     @Override
-    public Pagination<Payment> getPayments(final Long offset, final Long limit, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext context) {
+    public Pagination<Payment> getPayments(final Long offset, final Long limit, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext context) {
         return paymentProcessor.getPayments(offset, limit, withPluginInfo, properties, context, internalCallContextFactory.createInternalTenantContext(context));
```
Same here: The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -722,31 +722,31 @@ public Payment createChargebackReversal(final Account account, final UUID paymen
     @Override
-    public Pagination<Payment> getPayments(final Long offset, final Long limit, final String pluginName, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext tenantContext) throws PaymentApiException {
+    public Pagination<Payment> getPayments(final Long offset, final Long limit, final String pluginName, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext tenantContext) throws PaymentApiException {
         return paymentProcessor.getPayments(offset, limit, pluginName, withPluginInfo, properties, tenantContext, internalCallContextFactory.createInternalTenantContext(tenantContext));
```
Same here: The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -722,31 +722,31 @@ public Payment createChargebackReversal(final Account account, final UUID paymen
     @Override
-    public Payment getPaymentByExternalKey(final String paymentExternalKey, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext tenantContext)
+    public Payment getPaymentByExternalKey(final String paymentExternalKey, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext tenantContext)
             throws PaymentApiException {
```
Same here: The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -756,12 +756,12 @@ public Payment getPaymentByExternalKey(final String paymentExternalKey, final bo
     @Override
-    public Pagination<Payment> searchPayments(final String searchKey, final Long offset, final Long limit, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext context) {
+    public Pagination<Payment> searchPayments(final String searchKey, final Long offset, final Long limit, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext context) {
         return paymentProcessor.searchPayments(searchKey, offset, limit, withPluginInfo, properties, context, internalCallContextFactory.createInternalTenantContext(context));
```
Same here: The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented on payment/src/main/java/org/killbill/billing/payment/api/DefaultPaymentApi.java:
```diff
@@ -756,12 +756,12 @@ public Payment getPaymentByExternalKey(final String paymentExternalKey, final bo
     @Override
-    public Pagination<Payment> searchPayments(final String searchKey, final Long offset, final Long limit, final String pluginName, final boolean withPluginInfo, final Iterable<PluginProperty> properties, final TenantContext context) throws PaymentApiException {
+    public Pagination<Payment> searchPayments(final String searchKey, final Long offset, final Long limit, final String pluginName, final boolean withPluginInfo, final boolean withAttempts, final Iterable<PluginProperty> properties, final TenantContext context) throws PaymentApiException {
         return paymentProcessor.searchPayments(searchKey, offset, limit, pluginName, withPluginInfo, properties, context, internalCallContextFactory.createInternalTenantContext(context));
```
Same here: The new flag `withAttempts` is not passed (used) the  `paymentProcessor#getAccountPayments`

sbrossie commented:
@javier-gomez-hs I started to merge your PR but i realized you are missing some code -- i just commented. Could you add the logic to handle the new flag `withAttempts` and write associated tests to verify this works?

javier-gomez-hs commented:
@sbrossie 
I've added the missing property and fixed some Integration Tests.
Do you want me to add new Integration Tests for the PaymentApi and PaymentProcessor methods?

Thanks!

sbrossie commented:
@javier-gomez-hs Maybe to keep it simple what you could do is enhance your `testWithFailedPaymentAndScheduledAttempts` to also retrieve the payments using various payment apis (with the attempts). That would also give us confidence java-client is correctly handing attempts for various payment apis. Does that make sense?

sbrossie commented:
@javier-gomez-hs I started the merge on `work-for-release-0.17.x` but i end up with lots of conflicts because `master` and `work-for-release-0.17.x` have diverged quite a bit (this is risky and time consuming). 

Instead i suggest we close the PR and you resubmit against branch `work-for-release-0.17.x` (don't try to merge your branch into `work-for-release-0.17.x`  as you will see the same conflicts as i do. Instead checkout `work-for-release-0.17.x`  and port your diff into this branch and then submit your PR). 

We are having a lot of changes from different parts, so this is getting a bit tricky on our end (for next PR, i would suggest we always confirm beforehand against which branch this should be submitted to avoid this confusion). Let me know if this is not clear.

sbrossie commented:
Closed as work as been ported in PR #576
